Nodejs/Product/ProjectWizard/NodejsPackageParametersExtension.cs

```diff
@@ -14,23 +14,36 @@
 //
 //*********************************************************//
 
+using System;
 using System.Collections.Generic;
 using System.Text.RegularExpressions;
 using EnvDTE;
 using Microsoft.VisualStudio.TemplateWizard;
-
+
 namespace Microsoft.NodejsTools.ProjectWizard {
     class NodejsPackageParametersExtension : IWizard {
-        public void RunStarted(object automationObject, Dictionary<string, string> replacementsDictionary, WizardRunKind runKind, object[] customParams) {
-            var projectName = replacementsDictionary["$projectname$"];
+        private const int NpmPackageNameMaxLength = 214;
 
-            // Remove all leading url-invalid, underscore, and period characters from the string
+        /// <summary>
+        /// Normalize a project name to be a valid Npm package name.
+        /// </summary>
+        /// <param name="projectName">Name of a VS project.</param>
+        private string GetNpmSafeProjectName(string projectName) {
+            // Remove all leading url-invalid, underscore, and period characters
             var npmProjectNameTransform = Regex.Replace(projectName, "^[^a-zA-Z0-9-~]*", string.Empty);
 
             // Replace all invalid characters with a dash
-            npmProjectNameTransform = Regex.Replace(npmProjectNameTransform, "[^a-zA-Z0-9-_~.]", "-");
+            npmProjectNameTransform = Regex.Replace(npmProjectNameTransform, "[^a-zA-Z0-9-_~.]", "-");
+
+            // insert hypens between camelcased sections.
+            npmProjectNameTransform = Regex.Replace(npmProjectNameTransform, "([a-z0-9])([A-Z])", "$1-$2").ToLowerInvariant();
+
+            return npmProjectNameTransform.Substring(0, Math.Min(npmProjectNameTransform.Length, NpmPackageNameMaxLength));
+        }
```

Comment:
nit: move below public methods so that it's easier to understand the interfaces.